blockinfile - fix line ending regression

samdoran opened the pull request:
##### SUMMARY
<!--- Describe the change below, including rationale and design decisions -->
PR #66461 introduced a regression that resulted in an incorrect block in the file if the block to be inserted did not end with a line separator. This was causing some other tests in our CI to fail and probably would badly break things for others.

Fix this bug by ensuring the block has a line separator at the end and add tests to cover this scenario.
<!--- HINT: Include "Fixes #nnn" if you are fixing an existing issue -->

##### ISSUE TYPE
<!--- Pick one below and delete the rest -->
- Bugfix Pull Request
##### COMPONENT NAME
<!--- Write the short name of the module, plugin, task or feature below -->
`lib/ansible/modules/blockinfile.py`

samdoran committed:
blockinfile - fix line end regression

PR #66461 introduced a regression that resulted in an in correct block in the file
if the block to be inserted did not end with a line separator. Fix this bug and add
tests to cover this scenario.

samdoran commented:
Could you take a look, @riyad?

riyad commented:
I commented out your fix, but the tests still pass. It seems the tests don't quite capture the essence of the bug.

samdoran commented:
Hmmm, that's no good. I am able to see the bug and get the tests to fail by removing the patch.

Do you have a more complete solution? Otherwise we will need to revert this. 😞

samdoran commented:
Without the patch, I see the test failure:
```
> ansible-test integration --docker centos8 blockinfile -v

...

TASK [blockinfile : Add block without trailing line separator] *****************
changed: [testhost] => {"changed": true, "msg": "File created"}

TASK [blockinfile : Add block without trailing line separator again] ***********
changed: [testhost] => {"changed": true, "msg": "Block inserted"}

TASK [blockinfile : Check output file] *****************************************
ok: [testhost] => {"changed": false, "stat": {"atime": 1590611231.867382, "attr_flags": "", "attributes": [], "block_size": 4096, "blocks": 8, "charset": "us-ascii", "checksum": "597faeef3de556fa8f8769f75f991d259621aa36", "ctime": 1590611231.867382, "dev": 112, "device_type": 0, "executable": false, "exists": true, "gid": 0, "gr_name": "root", "inode": 3689547, "isblk": false, "ischr": false, "isdir": false, "isfifo": false, "isgid": false, "islnk": false, "isreg": true, "issock": false, "isuid": false, "mimetype": "text/plain", "mode": "0644", "mtime": 1590611231.867382, "nlink": 1, "path": "/root/ansible/test/results/.tmp/output_dir/test_blockinfile/chomped_block_test.txt", "pw_name": "root", "readable": true, "rgrp": true, "roth": true, "rusr": true, "size": 142, "uid": 0, "version": null, "wgrp": false, "woth": false, "writeable": true, "wusr": true, "xgrp": false, "xoth": false, "xusr": false}}

TASK [blockinfile : Ensure chomptest results are correct] **********************
fatal: [testhost]: FAILED! => {
    "assertion": "chomptest2 is not changed",
    "changed": false,
    "evaluated_to": false,
    "msg": "Assertion failed"
}

PLAY RECAP *********************************************************************
testhost                   : ok=23   changed=11   unreachable=0    failed=1    skipped=0    rescued=0    ignored=0

NOTICE: To resume at this test target, use the option: --start-at blockinfile
ERROR: Command "ansible-playbook blockinfile-o690h1qp.yml -i inventory -v" returned exit status 2.
Run command: docker exec 72edccd289f8a2d8aab9d88fc6543955be8c5f9d2a35e77f6b70c8202ea7416f tar czf /root/results.tgz --exclude  ...
Run command: docker exec -i 72edccd289f8a2d8aab9d88fc6543955be8c5f9d2a35e77f6b70c8202ea7416f dd if=/root/results.tgz bs=65536
Run command: tar oxzf /tmp/ansible-result-ml_54m_i.tgz -C /Users/sdoran/Source/ansible/test
Run command: docker rm -f 72edccd289f8a2d8aab9d88fc6543955be8c5f9d2a35e77f6b70c8202ea7416f
ERROR: Command "docker exec 72edccd289f8a2d8aab9d88fc6543955be8c5f9d2a35e77f6b70c8202ea7416f /usr/bin/env ANSIBLE_TEST_CONTENT_ROOT=/root/ansible LC_ALL=en_US.UTF-8 /usr/bin/python3.6 /root/ansible/bin/ansible-test integration blockinfile -v --metadata test/results/.tmp/metadata-q8hpw31a.json --truncate 131 --redact --color yes --requirements --allow-destructive" returned exit status 1.
```

samdoran commented:
And I get this in the file when running the test task without the fix, which is what was breaking our CI:
```
# BEGIN ANSIBLE MANAGED BLOCK
one
two
three# END ANSIBLE MANAGED BLOCK
# BEGIN ANSIBLE MANAGED BLOCK
one
two
three# END ANSIBLE MANAGED BLOCK
# BEGIN ANSIBLE MANAGED BLOCK
one
two
three# END ANSIBLE MANAGED BLOCK
```

riyad commented:
Weird ... seems to have been something wrong with my test setup.
I can confirm test capture the bug and your code fixes it. :+1: 
Sorry for the confusion.

riyad commented:
I'm getting a whole new appreciation for the work and rigor that is required to make Ansible reliable and well-behaved in all the (weird) circumstances. :trophy:

samdoran commented:
No worries. Thank you for confirming the fix. Much appreciated.
